### ISSUE TITLE
Is it possible not to report unchanged metrics?

### ISSUE THREAD

**Issue by czd890** [AppMetrics/AppMetrics#609]
```C#
class GaugeOptions{
ReportNoChange{get;set;}=true,
}
```
If the metrics is not updated, is it possible to add options without reporting.

or add custom filter ?

**Comment by alhardy** [comment]
What are you wanting to achieve?

**Comment by pwiens** [comment]
@alhardy I can't speak for @czd890 but we have the same need. Our problem is that we have a requirement to track some user specific information as a label for in our case we use counters and based on let's say userid we increment the counter with label userid=someguid. We also have our counters reset on reporting. I know this goes against everything I've read for tracking metrics in Prometheus and other time series databases but this is a requirement handed to us that we can't work around. We have a custom monitoring system that isn't based on Prometheus and apparently it's fine with high cardinality. I've written a custom reporter to send our metrics to this system but what ends up happening is the we see our memory usage growing slowly because of all of the label permutations. Ideally there would be a way to either have ResetOnReporting actually reset the concurrent dictionary or make it a little smarter and have it delete a specific permutation after x consecutive 0 value reports. If that's not something you want to entertain it would also be nice to be able to inject a custom counter. I'd be happy to submit a PR with either of these solutions.

**Comment by czd890** [comment]
@pwiens Yes, our needs are similar to yours

**Comment by alhardy** [comment]
@pwiens would be happy to take a PR which on resetting clears the dictionary rather than just resetting values when `ResetOnReporting` is set

**Comment by alhardy** [comment]
For very high cardinality such as tracking metrics based on userids for example as mentioned, AppMetrics isn't going to be a good solution. In saying that, I think generally I wouldn't use such high cardinality in a metrics monitoring solution, I'm not sure how this would be useful in terms of observing times series in something like grafana. Typically such user tracking would be done with something like google analytics. I'm not 100% sure, but with very high cardinality on a single metric you may run into issues with the underlying TSDB handling this also

**Comment by alhardy** [comment]
As per the assumption in my last comment see the [prometheus docs](https://prometheus.io/docs/practices/naming/) for example around high cardinality metrics.

> CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

**Comment by pwiens** [comment]
@alhardy I completely agree with you. I know what the correct approach would be for our use case but unfortunately the requirement to track these sort of metrics isn't something that I can get around I'm trying to find a non-terrible solution. Our metrics backend doesn't behave like a TSDB so that's not a concern for us. I have a custom reporter which sends the metrics to our backend. Btw, got started on this implementation but it turned out to be a bit more involved than I first thought then I got pulled into other stuff so I haven't had time to work on it. I'll try to pick it up again after things calm down.

**Comment by czd890** [comment]
Hi @pwiens @alhardy 😅 Are there any updates?

**Comment by justinmchase** [comment]
I'm not sure I fully understand but it sounds like you may want a logging solution instead. Maybe try Loki + Promtail? That would allow you to have specific events and you can aggregate those as datapoints in a time series more discretely.

**Comment by czd890** [comment]
Hi @justinmchase , what I want to say is: in our program, some metrics are not updated frequently, can we add an option to control when metrics are not updated, they will not be pushed to the database.

**Comment by EugeneKrapivin** [comment]
actually we have something resembling, we have some high cardinality metrics, however they are not reported constantly. At times there are series that keep reporting 0s (or infinetly decreasing values) even when a metrics wasn't pushed to the series (x Tags) for quite sometime. this hurts us because over time we can get up to 18K metrics a sec reported, while most of them being empty, requiring a restart to the services to clean out series that weren't recently updated but are still reported (using influx)

**Comment by EugeneKrapivin** [comment]
any updates on the matter?
